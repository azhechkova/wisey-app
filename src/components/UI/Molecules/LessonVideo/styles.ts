import { makeStyles } from 'tss-react/mui';

const useStyles = makeStyles()(() => ({
  video: {
    borderRadius: '1rem',
    width: '100%',
    height: 'calc(100% - 36px)',
    '@media (max-width: 992px)': {
      height: 'calc(100% - 82px)',
    },
  },
  root: {
    height: '100%',
  },
  info: {
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'center',
    '@media (max-width: 992px)': {
      flexDirection: 'column',
      gap: '10px',
      alignItems: 'flex-start',
    },
  },
  infoText: {
    padding: '6px 8px',
  },
}));

export default useStyles;
