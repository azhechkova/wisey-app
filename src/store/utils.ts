import { VIDEOS_STORAGE } from '~/constants';
import { TrackVideoType } from '~/types';
import { RootState } from './types';

export const getActiveLesson = (store: RootState): string =>
  store.app.activeLesson;

export const getVideoProgress = (store: RootState): TrackVideoType[] =>
  store.app.videoProgress;

export const loadState = (): Partial<RootState> | undefined => {
  try {
    const serializedState = localStorage.getItem(VIDEOS_STORAGE);

    if (serializedState === null) {
      return undefined;
    }

    return {
      app: {
        activeLesson: '',
        videoProgress: JSON.parse(serializedState),
      },
    };
  } catch (err) {
    return undefined;
  }
};

type SaveStateType = {
  videoProgress: TrackVideoType[];
};

export const saveState = (state: SaveStateType): void => {
  const serializedState = JSON.stringify(state.videoProgress);

  localStorage.setItem(VIDEOS_STORAGE, serializedState);
};
