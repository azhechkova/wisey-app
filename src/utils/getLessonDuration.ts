const getLabel = (hrs: number): string => {
  if (hrs) return 'hrs';

  return 'min';
};

const getLessonDuration = (duration: number): string => {
  const hours = Math.floor(duration / 3600);
  const minutes = Math.floor((duration - hours * 3600) / 60);
  const seconds = duration - hours * 3600 - minutes * 60;

  const intl = Intl.NumberFormat('en-US', { minimumIntegerDigits: 1 });
  const intlPadded = Intl.NumberFormat('en-US', {
    minimumIntegerDigits: 2,
  });

  const timeFormat = [hours, minutes, seconds]
    .map(val => {
      return val < 10 ? intlPadded.format(val) : intl.format(val);
    })
    .filter((val, i) => {
      if (i === 0) {
        return !(val === '00' || val === '0');
      }

      return true;
    })
    .join(':');

  const label = getLabel(hours);

  return `${timeFormat} ${label}`;
};

export default getLessonDuration;
