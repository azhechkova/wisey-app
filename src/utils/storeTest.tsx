import React, { PropsWithChildren } from 'react';
import { render } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import type { PreloadedState } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';

import {
  ToolkitStore,
  EnhancedStore,
} from '@reduxjs/toolkit/dist/configureStore';
import type { RootState } from '~/store/types';
import appReducer from '~/store/reducers/app';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>;
  store?: EnhancedStore<RootState>;
}

const initialState = {
  app: {
    activeLesson: '',
    videoProgress: [],
  },
};
const setupStore = (
  preloadedState?: PreloadedState<RootState>,
): ToolkitStore<RootState> => {
  return configureStore({
    reducer: { app: appReducer },
    preloadedState,
  });
};

export type AppStore = ReturnType<typeof setupStore>;

export const renderWithProviders = (
  ui: React.ReactElement,
  {
    preloadedState = initialState,
    store = configureStore({ reducer: { app: appReducer }, preloadedState }),
    ...renderOptions
  }: ExtendedRenderOptions = {},
): any => {
  const Wrapper = ({ children }: PropsWithChildren): JSX.Element => {
    return <Provider store={store}>{children}</Provider>;
  };

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
};
