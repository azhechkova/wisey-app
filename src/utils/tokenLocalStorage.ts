import constants from '../constants';

const setStorageToken = (token: string): void => {
  localStorage.setItem(constants.PROJECT_NAME, token);
};

const getStorageToken = (): string | null => {
  return localStorage.getItem(constants.PROJECT_NAME);
};

const removeStorageToken = (): void => {
  return localStorage.removeItem(constants.PROJECT_NAME);
};

export { setStorageToken, getStorageToken, removeStorageToken };
