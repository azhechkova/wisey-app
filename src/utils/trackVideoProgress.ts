import { VIDEOS_STORAGE } from '~/constants';
import { TrackVideoType } from '~/types';

export const trackVideoProgress = (trackVideo: TrackVideoType): void => {
  const videos = localStorage.getItem(VIDEOS_STORAGE);

  if (!videos) {
    localStorage.setItem(VIDEOS_STORAGE, JSON.stringify([trackVideo]));

    return;
  }

  const parsedVideos = JSON.parse(videos) as TrackVideoType[];

  if (!parsedVideos.some(video => video.videoId === trackVideo.videoId)) {
    const newProgress = parsedVideos.push(trackVideo);

    localStorage.setItem(VIDEOS_STORAGE, JSON.stringify(newProgress));
  }

  const newProgress = parsedVideos.map(video => {
    if (video.videoId === trackVideo.videoId) {
      return trackVideo;
    }

    return video;
  });

  localStorage.setItem(VIDEOS_STORAGE, JSON.stringify(newProgress));
};

export const getVideoProgress = (videoId: string): TrackVideoType => {
  const videos = localStorage.getItem(VIDEOS_STORAGE);
  const emptyProgress = { videoId, progress: 0, isFinished: false };

  if (!videos) {
    return emptyProgress;
  }

  const parsedVideos = JSON.parse(videos) as TrackVideoType[];
  const progressVideo = parsedVideos.find(video => videoId === video.videoId);

  if (!progressVideo) return emptyProgress;

  return progressVideo;
};
