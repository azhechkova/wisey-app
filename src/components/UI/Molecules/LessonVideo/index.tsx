import React, {
  SyntheticEvent,
  useCallback,
  useEffect,
  useRef,
  useState,
  VideoHTMLAttributes,
} from 'react';
import { Button, Typography, Box } from '@mui/material';
import CourseVideo from '~/components/UI/Atoms/CourseVideo';
import { useAppDispatch } from '~/store';
import { updateVideoProgress } from '~/store/reducers/app';
import {
  trackVideoProgress,
  getVideoProgress,
} from '~/utils/trackVideoProgress';
import { SPEED_KEYS } from '~/constants';
import useStyles from './styles';

interface LessonVideoProps extends VideoHTMLAttributes<HTMLVideoElement> {
  src?: string;
  duration: number;
}

const MIN_PLAYBACK_RATE = 0.25;
const MAX_PLAYBACK_RATE = 2;
const LessonVideo = ({
  src,
  duration,
  ...props
}: LessonVideoProps): JSX.Element => {
  const { classes } = useStyles();
  const dispatch = useAppDispatch();
  const ref = useRef<HTMLVideoElement | null>(null);
  const [isPipActive, setIsPipActive] = useState(false);

  const videoProgress = getVideoProgress(src || '');

  const onTimeUpdate = (event: SyntheticEvent<HTMLVideoElement>): void => {
    if (videoProgress?.isFinished) return;
    const target = event.target as HTMLVideoElement;
    const progress = target.currentTime;
    const isFinished = Math.round(progress) >= duration;

    const progressItem = {
      videoId: src || '',
      progress,
      isFinished,
    };

    dispatch(updateVideoProgress(progressItem));
    trackVideoProgress(progressItem);
  };

  const togglePiPMode = async (): Promise<void> => {
    const video = ref.current;

    if (!video) return;

    if (video !== document.pictureInPictureElement) {
      await video.requestPictureInPicture();
      setIsPipActive(true);
    } else {
      await document.exitPictureInPicture();
      setIsPipActive(false);
    }
  };

  const onPipOpen = async (): Promise<void> => {
    const isPiPSupported =
      'pictureInPictureEnabled' in document && document.pictureInPictureEnabled;

    if (!isPiPSupported) {
      alert('Picture in Picture is disabled in your browser.');

      return;
    }

    await togglePiPMode();
  };

  const increase = (): void => {
    const video = ref.current;

    if (!video) return;
    if (video.playbackRate < MAX_PLAYBACK_RATE) {
      video.playbackRate += MIN_PLAYBACK_RATE;
    }
  };

  const decrease = (): void => {
    const video = ref.current;

    if (!video) return;

    if (video.playbackRate > MIN_PLAYBACK_RATE) video.playbackRate -= 0.25;
  };

  const onKeyPress = useCallback((e: KeyboardEvent): void => {
    if (e.code === SPEED_KEYS.decrease.code) {
      decrease();
    } else if (e.code === SPEED_KEYS.increase.code) {
      increase();
    }
  }, []);

  useEffect(() => {
    window.addEventListener('keypress', onKeyPress);

    return () => {
      window.removeEventListener('keypress', onKeyPress);
    };
  }, [onKeyPress]);

  return (
    <Box className={classes.root}>
      <Box className={classes.info}>
        <Button onClick={onPipOpen}>
          {isPipActive ? 'Close video' : 'Open video in tab'}
        </Button>
        <Typography className={classes.infoText}>
          {SPEED_KEYS.decrease.label} - decrease speed;{' '}
          {SPEED_KEYS.increase.label} - increase speed
        </Typography>
      </Box>
      <CourseVideo
        onTimeUpdate={onTimeUpdate}
        controls
        src={src}
        currentTime={videoProgress?.progress || 0}
        className={classes.video}
        videoRef={ref}
        {...props}
      />
    </Box>
  );
};

export default LessonVideo;
