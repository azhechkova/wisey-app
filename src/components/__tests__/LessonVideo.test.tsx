import React from 'react';
import { fireEvent } from '@testing-library/react';
import { renderWithProviders } from '~/utils/storeTest';
import LessonVideo from '../UI/Molecules/LessonVideo';

const props = {
  src: 'https://example.com/video.mp4',
  duration: 120,
  id: 'video',
};

describe('LessonVideo', () => {
  it('renders properly', () => {
    const { container } = renderWithProviders(<LessonVideo {...props} />);

    expect(container).toMatchSnapshot();
  });

  it('should render the video player with right src', () => {
    const { container } = renderWithProviders(<LessonVideo {...props} />);

    const videoElement = container.querySelector('#video');

    expect(videoElement).toBeInTheDocument();
    expect(videoElement).toHaveAttribute('src', props.src);
  });

  it('should update the video progress and track the progress when the video is played', () => {
    const currentTime = 30;
    const event = {
      target: {
        currentTime,
      },
    };

    const { container } = renderWithProviders(<LessonVideo {...props} />);

    const videoElement = container.querySelector('#video');

    expect(videoElement).toBeInTheDocument();

    if (!videoElement) return;
    fireEvent.timeUpdate(videoElement, event);

    expect(localStorage.setItem).toHaveBeenCalledTimes(1);
  });

  it('should increase the playback speed when the increase key is pressed', () => {
    const { container } = renderWithProviders(<LessonVideo {...props} />);

    const videoElement = container.querySelector('#video') as HTMLVideoElement;

    fireEvent.keyPress(window, { code: 'KeyS' });
    expect(videoElement).toBeInTheDocument();

    expect(videoElement!.playbackRate).toBe(1.25);
  });

  it('should not increase the playback speed when the increase key is pressed and the current rate is the maximum', () => {
    const { container } = renderWithProviders(<LessonVideo {...props} />);

    const videoElement = container.querySelector('#video') as HTMLVideoElement;

    expect(videoElement).toBeInTheDocument();

    fireEvent.keyPress(window, { code: 'KeyS' });
    fireEvent.keyPress(window, { code: 'KeyS' });
    fireEvent.keyPress(window, { code: 'KeyS' });
    fireEvent.keyPress(window, { code: 'KeyS' });
    fireEvent.keyPress(window, { code: 'KeyS' });

    expect(videoElement.playbackRate).toBe(2);
    fireEvent.keyPress(window, { code: 'KeyS' });

    expect(videoElement.playbackRate).toBe(2);
  });
});
