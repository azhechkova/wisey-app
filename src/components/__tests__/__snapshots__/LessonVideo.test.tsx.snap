// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LessonVideo renders properly 1`] = `
<div>
  <div
    class="MuiBox-root css-3vegfa-root"
  >
    <div
      class="MuiBox-root css-1xwdhek-info"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Open video in tab
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <p
        class="MuiTypography-root MuiTypography-body1 css-q87npz-MuiTypography-root-infoText"
      >
        A
         - decrease speed;
         
        S
         - increase speed
      </p>
    </div>
    <video
      class="css-kptru5-video"
      controls=""
      id="video"
      src="https://example.com/video.mp4"
    >
      <track
        kind="captions"
      />
    </video>
  </div>
</div>
`;
